packages/SwingSet/src/kernel/vatManager/vat-warehouse.js

```diff
@@ -0,0 +1,257 @@
+   */
+  const ephemeral = {
+    /** @type {Map<string, VatInfo> } key is vatID */
+    vats: new Map(),
+  };
+
+  /** @type {Map<string, VatTranslators> } */
+  const xlate = new Map();
+  /** @param { string } vatID */
+  function findOrCreateTranslators(vatID) {
+    let translators = xlate.get(vatID);
+    if (!translators) {
+      // @@@ document pre-condition makeVatTranslators needs vatKeeper?
+      // or pass in the vatKeeper too;
+      // grep for getVatKeeper and avoid this pattern.
+      // remove ephemeral table from kernelKeeper@@@
+      translators = makeVatTranslators(vatID, kernelKeeper);
+      xlate.set(vatID, translators);
+    }
+    return translators;
+  }
+
+  /**
+   * @param {string} vatID
+   * @param {boolean} recreate
+   * @returns { Promise<VatInfo> }
+   */
+  async function ensureVatOnline(vatID, recreate) {
+    const info = ephemeral.vats.get(vatID);
+    if (info) return info;
+
+    // @@@getVatKeeper -> provideVatKeeper?
+    const vatKeeper =
+      kernelKeeper.getVatKeeper(vatID) || kernelKeeper.allocateVatKeeper(vatID);
+    const { source, options } = vatKeeper.getSourceAndOptions(); // @@@comms vat source repeated?
+
+    const translators = findOrCreateTranslators(vatID);
+
+    const createVat =
+      (!recreate && vatLoader.createVatDynamically) ||
+      (kernelKeeper.getDynamicVats().includes(vatID)
+        ? vatLoader.recreateDynamicVat
+        : vatLoader.recreateStaticVat);
+    // console.log('provide: creating from bundle', vatID);
+    const manager = await createVat(vatID, source, translators, options);
+
+    const { enablePipelining = false } = options;
+
+    // TODO: load from snapshot
+    await manager.replayTranscript();
+    const result = {
+      manager,
+      translators,
+      enablePipelining,
+    };
+    ephemeral.vats.set(vatID, result);
+    return result;
+  }
+
+  /**
+   * Bring new dynamic vat online and run its (bootstrap) code.
+   *
+   * @param {string} vatID
+   */
+  async function createDynamicVat(vatID) {
+    return ensureVatOnline(vatID, false);
+  }
+
+  /** @param { typeof console.log } logStartup */
+  async function start(logStartup) {
+    const recreate = true; // note: PANIC on failure to recreate
+
+    // instantiate all static vats
+    for (const [name, vatID] of kernelKeeper.getStaticVats()) {
+      logStartup(`allocateVatKeeper for vat ${name} as vat ${vatID}`);
+      // eslint-disable-next-line no-await-in-loop
+      await ensureVatOnline(vatID, recreate);
+    }
+
+    // instantiate all dynamic vats
+    for (const vatID of kernelKeeper.getDynamicVats()) {
+      logStartup(`allocateVatKeeper for dynamic vat ${vatID}`);
+      // eslint-disable-next-line no-await-in-loop
+      await ensureVatOnline(vatID, recreate);
+    }
+  }
+
+  /**
+   * @param { string } vatID
+   * @returns {{ enablePipelining?: boolean }
+   *  | void // undefined if the vat is dead or never initialized
+   * }
+   */
+  const lookup = vatID => {
```

Comment:
I'm curious, why use an arrow function here, but `function` elsewhere? I'm personally in favor of `function`, but I'm trying to get more comfortable with the `=>` style and I'm wondering if it's more appropriate in different situations.. maybe something that doesn't have side-effects?

Reply:
no good reason. will change.

Perhaps it started as something with a one-expression body; those look nicer as arrow functions, to my eye.